src/diff.c

```diff
@@ -426,81 +426,39 @@ static void strip_spaces(git_buf *buf)
 	git_buf_truncate(buf, len);
 }
 
-static int file_cb(
+int git_diff_patchid_print_callback__to_buf(
 	const git_diff_delta *delta,
-	float progress,
+	const git_diff_hunk *hunk,
+	const git_diff_line *line,
 	void *payload)
 {
 	struct patch_id_args *args = (struct patch_id_args *) payload;
 	git_buf buf = GIT_BUF_INIT;
-	int error;
+	int error = 0;
 
-	GIT_UNUSED(progress);
+	git_buf_sanitize(&buf);
 
-	if (!args->first_file &&
-	    (error = flush_hunk(&args->result, &args->ctx)) < 0)
+	if (line->origin == GIT_DIFF_LINE_CONTEXT_EOFNL ||
+	    line->origin == GIT_DIFF_LINE_ADD_EOFNL ||
+	    line->origin == GIT_DIFF_LINE_DEL_EOFNL)
 		goto out;
-	args->first_file = 0;
-
-	if ((error = git_buf_printf(&buf,
-				    "diff--gita/%sb/%s---a/%s+++b/%s",
-				    delta->old_file.path,
-				    delta->new_file.path,
-				    delta->old_file.path,
-				    delta->new_file.path)) < 0)
-		goto out;
-
-	strip_spaces(&buf);
 
-	if ((error = git_hash_update(&args->ctx, buf.ptr, buf.size)) < 0)
+	if ((error = (git_diff_print_callback__to_buf(delta, hunk,
+						      line, &buf))) < 0)
 		goto out;
 
-out:
-	git_buf_dispose(&buf);
-	return error;
-}
-
-static int patchid_line_cb(
-	const git_diff_delta *delta,
-	const git_diff_hunk *hunk,
-	const git_diff_line *line,
-	void *payload)
-{
-	struct patch_id_args *args = (struct patch_id_args *) payload;
-	git_buf buf = GIT_BUF_INIT;
-	int error;
-
-	GIT_UNUSED(delta);
-	GIT_UNUSED(hunk);
-
-	switch (line->origin) {
-	    case GIT_DIFF_LINE_ADDITION:
-		git_buf_putc(&buf, '+');
-		break;
-	    case GIT_DIFF_LINE_DELETION:
-		git_buf_putc(&buf, '-');
-		break;
-	    case GIT_DIFF_LINE_CONTEXT:
-		break;
-	    case GIT_DIFF_LINE_CONTEXT_EOFNL:
-	    case GIT_DIFF_LINE_ADD_EOFNL:
-	    case GIT_DIFF_LINE_DEL_EOFNL:
-		/*
-		 * Ignore EOF without newlines for patch IDs as whitespace is
-		 * not supposed to be significant.
-		 */
-		return 0;
-	    default:
-		git_error_set(GIT_ERROR_PATCH, "invalid line origin for patch");
-		return -1;
-	}
-
-	git_buf_put(&buf, line->content, line->content_len);
 	strip_spaces(&buf);
 
-	if ((error = git_hash_update(&args->ctx, buf.ptr, buf.size)) < 0)
+	if (line->origin == GIT_DIFF_LINE_FILE_HDR && !args->first_file)
+		if ((error = (flush_hunk(&args->result, &args->ctx))) < 0)
+			goto out;
```

Comment:
Stylistic nit: the nested if isn't required here and we usually use `&&` chaining in these cases. It's a matter of taste though, so feel free to ignore this comment.
Please remove the braces around `flush_hunk`, though.

Reply:
Ah good catch for the `&&`. This looks better.
I will remove all braces around functions calls.